client/cli/src/runner.rs

```diff
@@ -257,34 +199,18 @@ impl<C: SubstrateCli> Runner<C> {
 		}
 	}
 
-	fn run_service_until_exit<T, F>(mut self, service_builder: F) -> Result<()>
-	where
-		F: FnOnce(Configuration) -> std::result::Result<T, sc_service::error::Error>,
-		T: AbstractService + Unpin,
-	{
-		let service = service_builder(self.config)?;
-
-		// we eagerly drop the service so that the internal exit future is fired,
-		// but we need to keep holding a reference to the global telemetry guard
-		// and drop the runtime first.
-		let _telemetry = service.telemetry();
-
-		// we hold a reference to the base path so if the base path is a temporary directory it will
-		// not be deleted before the tokio runtime finish to clean up
-		let _base_path = service.base_path();
-
-		{
-			let f = service.fuse();
-			self.tokio_runtime
-				.block_on(main(f))
-				.map_err(|e| e.to_string())?;
-		}
-
-		// The `service` **must** have been destroyed here for the shutdown signal to propagate
-		// to all the tasks. Dropping `tokio_runtime` will block the thread until all tasks have
-		// shut down.
-		drop(self.tokio_runtime);
-
+	/// A helper function that runs a node with tokio and stops if the process receives the signal
+	/// `SIGTERM` or `SIGINT`.
+	pub fn run_node_until_exit(
+		mut self,
+		initialise: impl FnOnce(Configuration) -> sc_service::error::Result<TaskManager>,
+	) -> Result<()> {
+		self.print_node_infos();
+		let mut task_manager = initialise(self.config)?;
+		self.tokio_runtime.block_on(main(task_manager.future().fuse()))
+			.map_err(|e| e.to_string())?;
+		task_manager.terminate();
+		drop(task_manager);
```

Comment:
It's not the task manager that needs to be dropped for the tasks to finish but the tokio runtime.

This actually waits for all the blocking code in the tasks to finish before actually dropping.
```
drop(tokio_runtime)
```

Now this is not a big deal because:
```
drop(task_manager)
```
...does nothing (because we already terminate just before)

And the tokio_runtime is dropped at the end of the function anyway.

Reply:
Fixed in #6548